src/zip_code_tree.hpp

```diff
@@ -691,4 +872,758 @@ struct iterator_traits<vg::ZipCodeTree::reverse_iterator>{
+            }
+            
+
+        } else {
+        //If there is an open ancestor that isn't a chain, so the ancestor must be a snarl
+#ifdef DEBUG_ZIP_CODE_TREE
+            assert(forest_state.open_intervals.back().code_type == ZipCode::REGULAR_SNARL ||
+                   forest_state.open_intervals.back().code_type == ZipCode::IRREGULAR_SNARL ||
+                   forest_state.open_intervals.back().code_type == ZipCode::CYCLIC_SNARL ||
+                   forest_state.open_intervals.back().code_type == ZipCode::ROOT_SNARL);
+#endif
+
+            //Open the child chain
+            open_chain(forest_state, forest_state.open_intervals.size(), 
+                       forest_state.seed_sort_order[current_interval.interval_start], current_interval.is_reversed);
+            
+        }
+
+        if (current_interval.code_type != ZipCode::NODE) {
+            // Add to open_intervals
+            forest_state.open_intervals.emplace_back(std::move(current_interval));
+        }
+    }
+    //Finished adding all intervals
+    
+
+    //Now close anything that remained open
+    while (!forest_state.open_intervals.empty()) {
+        interval_and_orientation_t& ancestor_interval = forest_state.open_intervals.back();
+        const Seed& last_seed = seeds.at(forest_state.seed_sort_order[ancestor_interval.interval_end-1]);
+
+        if (ancestor_interval.code_type == ZipCode::CHAIN ||
+            ancestor_interval.code_type == ZipCode::ROOT_CHAIN ||
+            ancestor_interval.code_type == ZipCode::ROOT_NODE) {
+            //Close a chain
+
+            close_chain(forest_state, forest_state.open_intervals.size()-1, 
+                        last_seed, ancestor_interval.is_reversed); 
+        } else {
+#ifdef DEBUG_ZIP_CODE_TREE
+            assert(ancestor_interval.code_type == ZipCode::REGULAR_SNARL ||
+                   ancestor_interval.code_type == ZipCode::IRREGULAR_SNARL ||
+                   ancestor_interval.code_type == ZipCode::CYCLIC_SNARL ||
+                   ancestor_interval.code_type == ZipCode::ROOT_SNARL);
+#endif
+            //Close a snarl
+            close_snarl(forest_state, forest_state.open_intervals.size()-1, 
+                        last_seed, ancestor_interval.is_reversed, ancestor_interval.code_type == ZipCode::CYCLIC_SNARL); 
+        }
+
+        forest_state.open_intervals.pop_back();
+    }
+
+    if (trees[forest_state.active_zip_tree_i].zip_code_tree.size() == 0) {
+        trees.erase(trees.begin() + forest_state.active_zip_tree_i);
+    }
+#ifdef DEBUG_ZIP_CODE_TREE
+    print_self(&seeds);
+    validate_zip_forest(distance_index, &seeds, distance_limit);
+    assert(forest_state.open_chains.empty());
+    assert(forest_state.open_intervals.empty());
+#endif
+
+}
+
+template<typename Minimizer>
+vector<ZipCodeForest::interval_and_orientation_t> ZipCodeForest::get_cyclic_snarl_intervals(
```

Comment:
This is another one of those big algorithm functions we always write. Would it be possible to break any of the `/******* *******/` pieces out into their own functions to split this up?